### PR TITLE
Handing (?u)

### PR DESCRIPTION
**Pull request by mmatera**
(?u) is innecesary in 3.11. Add it conditionally to the version of the interpreter.

With this, just remains one error in the doctests.

### CODE REVIEW

File: mathics/eval/strings.py [hunk 1]
```diff
@@ -20,6 +21,27 @@
     "*": "*?",
 }
 
+if sys.version_info < (3, 11):
+    REGEX_UNICODE_PREFIX = r"(?u)"
+else:
+    REGEX_UNICODE_PREFIX = ""
+
+
+_SYMBOL_TO_REGEX_DICT = {
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/eval/strings.py:30]: Regarding if the dict is regenerated, is it still true if the items depends on a module variable?

[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/eval/strings.py:30]: Indeed, the conditional is not necessary. It seems `(?u)` was needed for the `2.*` versions.

File: mathics/builtin/string/regexp.py [hunk 2]
```diff
@@ -24,9 +24,9 @@ class RegularExpression(Builtin):
      = RegularExpression[[abc]]
 
     ## Mathematica doesn't seem to verify the correctness of regex
-    #> StringSplit["ab23c", RegularExpression["[0-9]++"]]
-     : Element RegularExpression[[0-9]++] is not a valid string or pattern element in RegularExpression[[0-9]++].
-     = StringSplit[ab23c, RegularExpression[[0-9]++]]
+    #> StringSplit["ab23c", RegularExpression["[0-9]++)"]]
```
Comment (rocky) [inline mathics/builtin/string/regexp.py:27]: Why these changes?

File: mathics/eval/strings.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import re
+import sys
```
Comment (rocky) [inline mathics/eval/strings.py:5]: Is `sys` used  all of a sudden below?

File: mathics/builtin/string/regexp.py (hunk 2 above)
Comment (mmatera) [reply mathics/builtin/string/regexp.py:27]: Because the previous pattern seems to be understood as a valid pattern by the new 
version of the regex parser

Comment (rocky) [reply mathics/builtin/string/regexp.py:27]: Ok. I guess we should fix that.

File: mathics/eval/strings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/eval/strings.py]: > And I answered because you were curious and to verify and substantiate what I wrote.

BTW, thank you for that.

> 
> My concern is that you are thinking about this kind of stuff and we spend time at these kinds of things rather than focusing on other areas include these areas that _could_ save us a significant amount of run time. In the past the code has been perverted a bit chasing down such minor things.

The aim was not to obtain a "micro speed-up", but to avoid a "micro slow-down" if the conditional was needed. 

> 
> A little knowledge is dangerous.

[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/eval/strings.py]: @rocky, in any case, I also had the feeling that putting that dictionary as module constant would be clearer than buried inside the routine, despite the possible overhead of using a module variable instead of a function variable. But again, I would prefer to leave that kind of decision to you.

**Comment by mmatera** [comment]
This line https://github.com/Mathics3/mathics-core/blame/0a967877a2f1fd47c1d79aa8b0117681cc252fac/mathics/eval/strings.py#L179  seems to be the responsable of the fails: it should return a second parameter `global_re_options`.

On the other hand, is it really any need for this second parameter?

**Comment by rocky** [comment]
> This line https://github.com/Mathics3/mathics-core/blame/0a967877a2f1fd47c1d79aa8b0117681cc252fac/mathics/eval/strings.py#L179 seems to be the responsable of the fails: it should return a second parameter `global_re_options`.
> 
> On the other hand, is it really any need for this second parameter?

No. In the next commit, all of this will be gone.  There is still a bit to do here.

**Comment by mmatera** [comment]
@rocky: OK, if there is something else I can do to help, just tell me.

**Comment by rocky** [comment]
If you'd like to move the regexp conversion code to  `mathics.core.convert.regexp` as you suggest that would be great! 

Or write some unit tests for `to_regex()` (rather than testing exclusively via `StringSplit()` and other builtin functions that would be great as well`

Or fill in the to_regex gaps that we have. I recall some TODOs here, that would be much appreciated as well.

**Comment by rocky** [comment]
Also, I think if we try this on 3.11 there are still some test failures.

**Comment by rocky** [comment]
And there is _also_ fixing the callers of `to_regexp` so that the look for failures (return None) and do something appropriate. 

I am always amazed at how if you look in detail at any single specific item like handling regular expressions, there is a ton things to do, even after straightening out the code a little bit more.
